Add "compare" on dataclass fields (Fixes: #8905)

simonschiele opened the pull request:
Add support for `compare` argument on dataclass field. This allows to conveniently remove single fields from getting used by dataclasses `__eq__` comparison.

### Description
Within annotated declarative setups for Models currently the `compare` argument is not supported for fields. Would be strongly appreciated to support this one, since it is a very convenient way to control the behavior of the generated `__eq__` for transformed dataclasses.

Change simply adds the parameter to the `_AttributeOptions` used for dataclass-generation, which allows for usage within `mapped_column()` and `relationship()` like for the other supported arguments.

Fixes #8905

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [X] A short code fix
	- fixes 
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

simonschiele committed:
Add "compare" on dataclass fields (Fixes: #8905)

Add support for `compare` argument on dataclass fields to remove single
fields from comparision by `__eq__`.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 3a9feb5dfc326705998fbba7f389ff08da62520a of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change 3a9feb5dfc326705998fbba7f389ff08da62520a: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4254

CaselIT commented on test/orm/declarative/test_dc_transforms.py:
```diff
@@ -565,6 +565,18 @@ class Engineer(MappedAsDataclass, Person):
             ),
         )
 
+    @testing.only_if(lambda: compat.py310, "python 3.10 is required")
```
I think we can remove this, since compare has always been an option of dataclass

simonschiele replied:
Changed, thanks.

CaselIT reviewed (commented):
Looks ok, just a minor change to the tests

simonschiele commented:
> hm I wonder why i didnt include "compare" in the first place?

I blame the naming. Due to the feature being controlled on class-level by `eq` and on field level by `compare` I also found this to be highly confusing.

In theory we are also missing `hash` and `metadata` on the field-level. Didn't add these to keep the scope reasonable. Also couldn't come up with a use-case for `hash` on the fly.

simonschiele committed:
Remove 3.10 fencing on dc-compare test

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision ea9a53d2ca60befdd0c570013c0e57a78c11dd4a of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset ea9a53d2ca60befdd0c570013c0e57a78c11dd4a added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4254

simonschiele committed:
Adjust associationproxy for dc-compare

simonschiele committed:
Add dh-compare param to docstrings

simonschiele commented:
> **mike bayer** (zzzeek) wrote:
> I just pushed [1057b47](https://github.com/sqlalchemy/sqlalchemy/commit/1057b47bca2522e45d9621a709d033aa4fb88888) which means you will need to rebase and add to associationproxy.py also

Thanks :-)

Rebased, adjusted and also added the parameter to the corresponding docstrings.

zzzeek commented:
oh thanks!  I just rebased on this end and added a few docstirngs, so we are out of sync by one commit.  It's all set, so thanks very much!!

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4254 has been **merged**. Congratulations! :)
